modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/AbstractUrlSigningProvider.java

```diff
@@ -101,24 +109,32 @@ public void setSecurityService(SecurityService securityService) {
    * @return The current set of url beginnings this signing provider is looking for.
    */
   public Set<String> getUris() {
-    return keys.values().stream()
-            .map(keyEntry -> keyEntry.url)
-            .collect(Collectors.collectingAndThen(
-                    Collectors.toSet(),
-                    Collections::unmodifiableSet));
+    return Collections.unmodifiableSet(urls.keySet());
   }
 
   /**
-   * If available get a {@link KeyEntry} if there is a matching Url matcher.
+   * If available get a {@link Key} if there is a matching Url matcher.
    *
    * @param baseUrl
    *          The url to check against the possible matchers.
-   * @return The {@link KeyEntry} if it is available.
+   * @return The {@link Key} if it is available.
    */
-  private Optional<Map.Entry<String, KeyEntry>> getKeyEntry(String baseUrl) {
-    return keys.entrySet().stream()
-            .filter(entry -> baseUrl.startsWith(entry.getValue().url))
-            .findAny();
+  private Key getKey(String baseUrl) {
+    /* Optimization: Use TreeMap.floorEntry that can retrieve the greatest URL equal to or greater than 'baseUrl'
+       in O(log(n)). As we are trying to find an URL that is a prefix of 'baseUrl', candidate.getKey() either is
+       that URL (needs to be checked!) or there is no such URL. */
+    Map.Entry<String, Key> candidate = urls.floorEntry(baseUrl);
+    if (candidate != null && baseUrl.startsWith(candidate.getKey())) {
+      Key key = candidate.getValue();
+
+      // Don't accept URLs without an organization context
```

Comment:
Depends on what semantics you assume. Here, the method returns a key that can be used to sign "baseUrl". Since the keys explicitly support organizations, a key that cannot be used by the current organization to sign "baseUrl" is therefore not returned.

I've added some information to the comment, though.

Reply:
Everything always depends on the semantics you assume. :wink: But okay, your point is fair, and I can live with a note in the docs. :slightly_smiling_face: